bioio_base/embedded_metadata.py

```diff
@@ -0,0 +1,136 @@
+from typing import (
+    Any,
+    ClassVar,
+    Generic,
+    Optional,
+    Sequence,
+    Set,
+    TypeVar,
+    Union,
+    overload,
+)
+
+from .dimensions import DimensionNames
+
+# -----------------------------------------------------------------------------
+# Descriptor for metadata fields.
+# -----------------------------------------------------------------------------
+T = TypeVar("T")  # Generic type variable to enable type-safe storage of metadata
+
+
+class MetadataField(Generic[T]):
+    def __init__(self, label: str, default: Optional[T] = None) -> None:
+        self.label = label
+        self.default = default
+        self.private_name: Optional[str] = None
+
+    def __set_name__(self, owner: Any, name: str) -> None:
+        self.private_name = "_" + name
+
+    @overload
+    def __get__(self, instance: None, owner: Any) -> "MetadataField[T]":
+        ...
+
+    @overload
+    def __get__(self, instance: Any, owner: Any) -> Optional[T]:
+        ...
+
+    def __get__(
+        self, instance: Any, owner: Any = None
+    ) -> Union["MetadataField[T]", Optional[T]]:
+        if instance is None:
+            return self  # When accessed on the class, return the descriptor.
+        return instance.__dict__.get(self.private_name, self.default)
+
+    def __set__(self, instance: Any, value: T) -> None:
+        instance.__dict__[self.private_name] = value
+
+
+# -----------------------------------------------------------------------------
+# EmbeddedMetadata class with definitive (closed) fields.
+# -----------------------------------------------------------------------------
+class EmbeddedMetadata:
+    # Calculated metadata fields.
+    dimensions_present: MetadataField[Optional[Sequence[str]]] = MetadataField(
+        "Dimensions Present"
+    )
+    image_size_c: MetadataField[Optional[int]] = MetadataField("Image Size C")
```

Comment:
Can you add descriptions to each of these? Descriptions for them can be found in LabKey

Reply:
resolved in [8d2c263](https://github.com/bioio-devs/bioio-base/pull/29/commits/8d2c263e2f17e05e292673dcacc3856be8aaa39d)